[Enhancement] Add optimizer rule to rewrite Left Join to Inner Join

decster opened the pull request:
This commit introduces a new optimizer transformation rule, `RewriteLeftJoinToInnerJoinRule`. This rule identifies `LEFT OUTER JOIN` operations that can be safely converted to `INNER JOIN` operations.

The conditions for this transformation are:
1. A foreign key constraint exists from the left table to the right table involved in the join.
2. All columns on the left table that are part of the foreign key constraint are defined as NOT NULL.
3. All columns participating in the foreign key constraint are part of an equality predicate in the join's ON condition.

This optimization can lead to more efficient query plans by allowing the optimizer to consider a wider range of join strategies and by enabling further optimizations like table pruning for the right-hand side of the original join.

The rule has been integrated into the logical optimization phase and includes unit tests to cover various scenarios, including those mentioned in issue #59101. This change is expected to improve performance for queries and materialized views that meet the specified criteria.

## Why I'm doing:

## What I'm doing:

Fixes #59101

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.5
  - [ ] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1

google-labs-jules[bot] committed:
feat: Add optimizer rule to rewrite Left Join to Inner Join

This commit introduces a new optimizer transformation rule,
`RewriteLeftJoinToInnerJoinRule`. This rule identifies
`LEFT OUTER JOIN` operations that can be safely converted to
`INNER JOIN` operations.

The conditions for this transformation are:
1. A foreign key constraint exists from the left table to the
   right table involved in the join.
2. All columns on the left table that are part of the foreign
   key constraint are defined as NOT NULL.
3. All columns participating in the foreign key constraint are
   part of an equality predicate in the join's ON condition.

This optimization can lead to more efficient query plans by
allowing the optimizer to consider a wider range of join
strategies and by enabling further optimizations like table pruning
for the right-hand side of the original join.

The rule has been integrated into the logical optimization phase
and includes unit tests to cover various scenarios, including
those mentioned in issue #59101. This change is expected to
improve performance for queries and materialized views that
meet the specified criteria.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=59179) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=59179) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=59179) it.</sub>

google-labs-jules[bot] committed:
feat: Add optimizer rule to rewrite Left Join to Inner Join

This commit introduces a new optimizer transformation rule,
`RewriteLeftJoinToInnerJoinRule`. This rule identifies
`LEFT OUTER JOIN` operations that can be safely converted to
`INNER JOIN` operations.

The conditions for this transformation are:
1. A foreign key constraint exists from the left table to the
   right table involved in the join.
2. All columns on the left table that are part of the foreign
   key constraint are defined as NOT NULL.
3. All columns participating in the foreign key constraint are
   part of an equality predicate in the join's ON condition.

This optimization can lead to more efficient query plans by
allowing the optimizer to consider a wider range of join
strategies and by enabling further optimizations like table pruning
for the right-hand side of the original join.

The rule has been integrated into the logical optimization phase
and includes unit tests to cover various scenarios, including
those mentioned in issue #59101. This change is expected to
improve performance for queries and materialized views that
meet the specified criteria.

decster commented:
this is just a test for google jules, and the result is reaching merge standard, lot of compile errors and wrong import/parameter types. close this first
